### PR TITLE
cpu/sam0: improve ethernet driver resilience

### LINKED ISSUE #16451: sam0_eth: extremely long time to RX (frames stuck in buffer?)
#### Description
If `same54-xpro` is connected to a network with some random (broadcast?) network traffic, after a while it will get very slow to respond to incoming frames. 

#### Steps to reproduce the issue
Flash `examples/gnrc_networking` onto the `same54-xpro`

```
2021-05-05 21:49:59,921 # Iface  5  HWaddr: FC:C2:3D:0D:2D:1F 
2021-05-05 21:49:59,925 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2021-05-05 21:49:59,927 #           RTR_ADV  
2021-05-05 21:49:59,930 #           Source address length: 6
2021-05-05 21:49:59,932 #           Link type: wired
2021-05-05 21:49:59,938 #           inet6 addr: fe80::fec2:3dff:fe0d:2d1f  scope: link  VAL
2021-05-05 21:49:59,941 #           inet6 group: ff02::2
2021-05-05 21:49:59,943 #           inet6 group: ff02::1
2021-05-05 21:49:59,947 #           inet6 group: ff02::1:ff0d:2d1f
2021-05-05 21:49:59,948 #           
2021-05-05 21:49:59,951 #           Statistics for Layer 2
2021-05-05 21:49:59,955 #             RX packets 49273  bytes 5847245
2021-05-05 21:49:59,960 #             TX packets 18032 (Multicast: 15)  bytes 2129894
2021-05-05 21:49:59,963 #             TX succeeded 18032 errors 0
2021-05-05 21:49:59,966 #           Statistics for IPv6
2021-05-05 21:49:59,970 #             RX packets 18838  bytes 1962063
2021-05-05 21:49:59,975 #             TX packets 18032 (Multicast: 15)  bytes 1877446
2021-05-05 21:49:59,978 #             TX succeeded 18032 errors 0
```

ping the link-local address from a Linux machine

```
ping fe80::fec2:3dff:fe0d:2d1f%eno1
PING fe80::fec2:3dff:fe0d:2d1f%eno1(fe80::fec2:3dff:fe0d:2d1f%eno1) 56 data bytes
…
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=151 ttl=64 time=0.668 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=152 ttl=64 time=0.329 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=153 ttl=64 time=0.324 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=154 ttl=64 time=0.312 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=155 ttl=64 time=0.338 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=156 ttl=64 time=0.321 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=157 ttl=64 time=0.309 ms
```

Now flood the network with broadcast traffic

```
sudo ping -f ff02::1%eno1
```

#### Expected results

We might see some disruption, but the when ending the broadcast ping, everything should return to normal.

#### Actual results

The broadcast storm triggers an issue where frames get seemingly stuck in the RX buffer of the ethernet peripheral.
They are delivered to the applications with a delay of several seconds.

```
4 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=178 ttl=64 time=0.319 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=179 ttl=64 time=0.315 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=180 ttl=64 time=11.7 ms # broadcast storm starts
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=181 ttl=64 time=7.07 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=182 ttl=64 time=3.52 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=183 ttl=64 time=95.4 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=184 ttl=64 time=4091 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=185 ttl=64 time=4128 ms # broadcast storm ends
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=186 ttl=64 time=4070 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=188 ttl=64 time=3985 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=189 ttl=64 time=4032 ms
…
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=312 ttl=64 time=5081 ms # device never recovers
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=313 ttl=64 time=5955 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=314 ttl=64 time=5952 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=315 ttl=64 time=6246 ms
64 bytes from fe80::fec2:3dff:fe0d:2d1f%eno1: icmp_seq=316 ttl=64 time=6053 ms
```

packets sent from RIOT with `udp send` are delivered instantly, but packets sent with

    nc -u -6 fe80::fec2:3dff:fe0d:2d1f%eno1

take several seconds to be received.

#### Versions
RIOT master

### PR DESCRIPTION
**Pull request by dylad**
### Contribution description
This PR provides a better resilience for the SAM0 ethernet driver in case of heavy traffic in the network.

In some case interrupts fire faster than the netdev thread can process them and we run out of buffers quickly. With this PR, we now check if we don't have buffers available, so we can flush everything and restart reception properly even if we did drop a few in the operation.

### Testing procedure
See #16451 

### Issues/PRs references
Fixes #16451

### CODE REVIEW

**Comment by dylad** [comment]
@benpicco I need to do some cleanup with this PR but it should solve your issue.

File: cpu/sam0_common/periph/eth.c [hunk 1]
```diff
@@ -193,13 +204,27 @@ int sam0_eth_send(const struct iolist *iolist)
...
 {
     (void)block;
     unsigned rxlen = 0;
     uint16_t idx = rx_idx;
-    /* Ensure we are at the beginning of the new frame */
-    while (!(rx_curr->address & DESC_RX_ADDR_OWNSHP) && (rx_curr->status & DESC_RX_STATUS_STA_FRAME)) {}
+    uint8_t tmp = ETH_RX_BUFFER_COUNT;
+
+    do {
+        if ((rx_curr->address & DESC_RX_ADDR_OWNSHP) && (rx_curr->status & DESC_RX_STATUS_STA_FRAME)) {
```
Comment (benpicco) [inline cpu/sam0_common/periph/eth.c:215]:
```suggestion
        if ((rx_curr->address & DESC_RX_ADDR_OWNSHP) &&
            (rx_curr->status & DESC_RX_STATUS_STA_FRAME)) {
```

Comment (benpicco) [inline cpu/sam0_common/periph/eth.c:215]: Please add a comment what's going on here

File: cpu/sam0_common/periph/eth.c [hunk 2]
```diff
@@ -90,6 +90,17 @@ static uint8_t  rx_idx;
 
 static uint8_t  rx_buf[ETH_RX_BUFFER_COUNT][ETH_RX_BUFFER_SIZE] __attribute__((aligned(GMAC_BUF_ALIGNMENT)));
 static uint8_t  tx_buf[ETH_TX_BUFFER_COUNT][ETH_TX_BUFFER_SIZE] __attribute__((aligned(GMAC_BUF_ALIGNMENT)));
+extern sam0_eth_netdev_t _sam0_eth_dev;
+
+void sam0_clear_rx_buffers(void)
```
Comment (benpicco) [inline cpu/sam0_common/periph/eth.c:95]: better add a comment stating that this is called in interrupt context

File: cpu/sam0_common/sam0_eth/eth-netdev.c [hunk 3]
```diff
@@ -140,24 +141,24 @@ void sam0_eth_setup(netdev_t* netdev)
...
-    tsr = GMAC->TSR.reg;
     rsr = GMAC->RSR.reg;
+    (void)isr;
 
     if (rsr & GMAC_RSR_REC) {
         netdev_trigger_event_isr(_sam0_eth_dev.netdev);
     }
-
-    GMAC->TSR.reg = tsr;
+    if (rsr & GMAC_RSR_BNA) {
```
Comment (benpicco) [inline cpu/sam0_common/sam0_eth/eth-netdev.c:156]: Better add a comment as to what that flag means and when it occurs

**Review by benpicco — COMMENTED** [review]
Nice, thank you!

I still get the occasional 

```
2021-07-26 22:39:03,225 # gnrc_netif: possibly lost interrupt.
2021-07-26 22:39:03,229 # gnrc_netif: possibly lost interrupt.
2021-07-26 22:39:03,233 # gnrc_netif: possibly lost interrupt.
2021-07-26 22:39:03,237 # gnrc_netif: possibly lost interrupt.
```

but the interface remains functional. This makes the ethernet interface properly useful!

File: cpu/sam0_common/periph/eth.c (hunk 1 above)
Comment (dylad) [reply cpu/sam0_common/periph/eth.c]: Done.

Comment (dylad) [reply cpu/sam0_common/periph/eth.c]: Fixed.

File: cpu/sam0_common/periph/eth.c (hunk 2 above)
Comment (dylad) [reply cpu/sam0_common/periph/eth.c:97]: Done.

File: cpu/sam0_common/sam0_eth/eth-netdev.c (hunk 3 above)
Comment (dylad) [reply cpu/sam0_common/sam0_eth/eth-netdev.c:161]: Done.

File: cpu/sam0_common/periph/eth.c [hunk 4]
```diff
@@ -193,13 +206,33 @@ int sam0_eth_send(const struct iolist *iolist)
...
+       buffers due to BNA interrupt), return an error to netdev so
+       we can move forward */
+    do {
+        if ((rx_curr->address & DESC_RX_ADDR_OWNSHP)
+            && (rx_curr->status & DESC_RX_STATUS_STA_FRAME)) {
+            break;
+        }
+        else {
+            rx_idx = (rx_idx+1) % ETH_RX_BUFFER_COUNT;
+            rx_curr = &rx_desc[rx_idx];
```
Comment (benpicco) [inline cpu/sam0_common/periph/eth.c:228]: Do we really want to modify `rx_idx` here, not `idx`?

**Review by benpicco — APPROVED** [review]
Works well, please squash!

**Pull request merged**